File: output/gemma4-e4b/avatar/Python/Java/atcoder_ABC165_C.java
import java.util.Scanner;
import java.util.Arrays;

public class atcoder_ABC165_C {

    private static int maxScore = 0;
    private static int N, M, Q;
    private static int[] a, b, c, d;

    public static void solve(int N, int M, int Q, int[] a, int[] b, int[] c, int[] d) {
        Solution.N = N;
        Solution.M = M;
        Solution.Q = Q;
        Solution.a = a;
        Solution.b = b;
        Solution.c = c;
        Solution.d = d;
        
        maxScore = 0;
        
        // currentSequence stores the sequence A of length N
        int[] currentSequence = new int[N];
        
        // Start recursion from index 0
        rec(0, currentSequence);
        
        System.out.println(maxScore);
    }

    /**
     * Recursive function to generate all non-decreasing sequences of length N.
     * @param k The current index being filled (0 to N-1).
     * @param currentSequence The array storing the sequence being built.
     */
    private static void rec(int k, int[] currentSequence) {
        if (k == N) {
            // Base case: Sequence A is complete
            int currentScore = 0;
            for (int i = 0; i < Q; i++) {
                // Check condition: A[b[i] - 1] - A[a[i] - 1] == c[i]
                // Note: Input indices a[i] and b[i] are 1-based.
                // Array indices are 0-based.
                
                // Get values from the sequence A (currentSequence)
                int valB = currentSequence[b[i] - 1];
                int valA = currentSequence[a[i] - 1];
                
                if (valB - valA == c[i]) {
                    currentScore += d[i];
                }
            }
            maxScore = Math.max(maxScore, currentScore);
            return;
        }

        // Determine the starting value for the current element A[k]
        int startValue;
        if (k == 0) {
            startValue = 1;
        } else {
            // Must be >= the previous element
            startValue = currentSequence[k - 1];
        }

        // Iterate through possible values for A[k]
        for (int i = startValue; i <= M; i++) {
            currentSequence[k] = i;
            rec(k + 1, currentSequence);
        }
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        
        if (!scanner.hasNextInt()) {
            // Handle empty input case if necessary
            return;
        }

        // Read N, M, Q
        N = scanner.nextInt();
        M = scanner.nextInt();
        Q = scanner.nextInt();

        // Initialize arrays
        a = new int[Q];
        b = new int[Q];
        c = new int[Q];
        d = new int[Q];

        // Read Q sets of parameters
        for (int i = 0; i < Q; i++) {
            a[i] = scanner.nextInt();
            b[i] = scanner.nextInt();
            c[i] = scanner.nextInt();
            d[i] = scanner.nextInt();
        }
        
        scanner.close();

        solve(N, M, Q, a, b, c, d);
    }
}

File: output/gemma4-e4b/avatar/Java/Python/atcoder_ABC165_C.py
import sys

# Set recursion limit higher for deep DFS calls
# The maximum depth is N, which can be up to M.
sys.setrecursionlimit(2000)

# Global variables to store state
a = []
b = []
c = []
d = []
n = 0
m = 0
q = 0
ans = -float('inf')

def dfs(list_sequence):
    global ans
    global n, m, q, a, b, c, d

    # Base Case: Sequence length reached N
    if len(list_sequence) == n:
        score = 0
        for i in range(q):
            # Check the condition: list[b[i]] - list[a[i]] == c[i]
            # Note: a[i] and b[i] are 0-based indices already.
            if list_sequence[b[i]] - list_sequence[a[i]] == c[i]:
                score += d[i]
        ans = max(ans, score)
        return

    # Recursive Step
    if not list_sequence:
        # First element: range 1 to M
        start_num = 1
    else:
        # Subsequent elements: range from the last element to M (non-decreasing sequence)
        start_num = list_sequence[-1]

    for num in range(start_num, m + 1):
        # Optimization: If we already reached N, we don't need to proceed, 
        # but the base case handles this check implicitly.
        
        # Add number and recurse
        list_sequence.append(num)
        dfs(list_sequence)
        # Backtrack
        list_sequence.pop()

def solve():
    global n, m, q, a, b, c, d, ans
    
    # Read all input from stdin
    input_data = sys.stdin.read().split()
    if not input_data:
        return

    data_index = 0
    
    # Read N, M, Q
    n = int(input_data[data_index])
    m = int(input_data[data_index + 1])
    q = int(input_data[data_index + 2])
    data_index += 3

    # Initialize arrays
    a = [0] * q
    b = [0] * q
    c = [0] * q
    d = [0] * q

    # Read constraints
    for i in range(q):
        # a[i] and b[i] are 1-based in input, convert to 0-based
        a[i] = int(input_data[data_index]) - 1
        b[i] = int(input_data[data_index + 1]) - 1
        c[i] = int(input_data[data_index + 2])
        d[i] = int(input_data[data_index + 3])
        data_index += 4

    # Start DFS
    dfs([])

    # Output the result
    print(ans)

if __name__ == "__main__":
    solve()